include/deal.II/base/mpi.h

```diff
@@ -547,6 +564,37 @@ namespace Utilities
           }
       }
 
+      template <typename T>
+      inline
+      void all_reduce (const MPI_Op    &mpi_op,
+                       const Vector<T> &values,
+                       const MPI_Comm  &mpi_communicator,
+                       Vector<T>  &output)
+      {
+#ifdef DEAL_II_WITH_MPI
+        if (job_supports_mpi())
+          {
+            if (values.begin() != output.begin())
+              output.reinit (values.size());
+
+            MPI_Allreduce ((values.begin() != output.begin()
+                            ?
+                            const_cast<void *>(static_cast<const void *>(values.begin()))
+                            :
+                            MPI_IN_PLACE),
+                           output.begin(), values.size(), internal::mpi_type_id((T *)0), mpi_op,
+                           mpi_communicator);
+          }
+        else
+#endif
+          {
+            (void)mpi_op;
+            (void)mpi_communicator;
+            output = values;
+          }
+      }
+
```

Comment:
I think this would be a case to use a single function for all three types of the reduce function (fixed-size array, std::vector, dealii::Vector) with plain pointers to the data. No need to duplicate the same function three times. Since it's an `internal` method, we do not expose the plain pointer variant to the user (well, at least not in the non-internal namespace).